File: go/pythagorean-triplet/pythagorean_triplet.go
package pythagorean

import (
	"math"
)

// Triplet represents a Pythagorean Triangle
type Triplet [3]int

// Range returns a list of all Pythagorean triplets with sides in the
// range min to max inclusive.
func Range(min, max int) (triplets []Triplet) {
	for a := min; a <= max-3; a++ {
		for b := a + 1; b <= max-2; b++ {
			c := int(math.Sqrt(float64(a*a + b*b)))
			if c <= max && a*a+b*b == c*c {
				triplets = append(triplets, Triplet{a, b, c})
			}
		}
	}
	return
}

// Sum returns a list of all Pythagorean triplets where the sum a+b+c
// (the perimeter) is equal to p.
func Sum(p int) (triplets []Triplet) {
	for a := 1; a <= p/3; a++ {
		b := ((p-a)*(p-a) - a*a) / (2 * (p - a))
		c := p - a - b
		if a < b && a*a+b*b == c*c {
			triplets = append(triplets, Triplet{a, b, c})
		}
	}
	return
}

File: javascript/pythagorean-triplet/pythagorean-triplet.js
export class Triplet {
    constructor(...sides) {
        this.sides = sides;
    }

    sum() {
        return this.sides.reduce((a, b) => a + b);
    }

    product() {
        return this.sides.reduce((a, b) => a * b);
    }

    isPythagorean() {
        const [a, b, c] = this.sides;
        return c*c == a*a + b*b;
    }

    static where({ minFactor: min=3, maxFactor: max=6, sum: sum }) {
        let triplets = [];

        for (let a = min; a < max; a++) {
            for (let b = a+1; b < max; b++) {
                const c = parseInt(Math.sqrt(a*a + b*b));
                const triplet = new Triplet(a, b, c);
                if (triplet.isPythagorean() && (!sum || triplet.sum() == sum)) {
                    triplets.push(triplet);
                }
            }
        }

        return triplets;
    }
}
